### ISSUE TITLE
Consider faster copy constructors for System and Forces

### ISSUE THREAD

**Issue by jchodera** [pandegroup/openmm#542]
Current timings for `repex.testsystems` deep copy (via XML-based serialization):
```
SrcImplicit (4091 atoms)
Copying took 0.871 seconds

SrcExplicit (56038 atoms)
Copying took 2.789 seconds

LysozymeImplicit (2621 atoms)
Copying took 0.603 seconds
```
@peastman suggests testing copy through `SerializationNode` to compare timings and see if this might be an easy way to speed this up.

This is just a reminder for me to tackle this timing at some point.

**Comment by jchodera** [comment]
I finally had a chance to test the timing here.  For the `SrcExplicit` from `openmmtools`, here are the timings:
```
Loading system.xml...
Done.
Deep copy...
  40.778 s elapsed (   4.078 s/call)
SerlizationNode based copy...
  13.043 s elapsed (   1.304 s/call)
```
Timing code is [here](https://gist.github.com/jchodera/f87beab7c7a0f6caf867).

Your suggestion of using a `SerializationNode` copy constructor seems to be significantly faster!  Would be great to have this implemented at some point.

**Comment by jchodera** [comment]
Bumping this. My tests found this would provide a significant speedup to us in an area of the code that frustrates debugging.  Would be great if this could be added!

**Comment by jchodera** [comment]
Larger systems seem to be incredibly, horrifically slow.
I used the XML files for project 10492, which we are actually trying to run on FAH.  
The deep copy takes an immensely frustrating 19.8 s; the SerializationNode scheme takes only 6.8 s.  This really adds up when we have to do it many times to generate alchemical intermediates.

Any chance we could get this feature into the 6.3 release?

**Comment by jchodera** [comment]
Even better would be to have real Copy constructors someday....

**Comment by peastman** [comment]
I don't think this is safe to add in 6.3, now that we're well past the beta release.  There are some complicated issues in how to implement it.  I'll try to do it soon after 6.3 goes out.

Since you took eight months to get around to testing it, I don't think you can claim it's *that* high of a priority. :)

**Comment by jchodera** [comment]
It took me 8 months to test it because it required remembering how the
crappy C++ build system works.

**Comment by jchodera** [comment]
I'll note that I have been arguing for real copy constructors since at
least before I left Stanford in 2008.  I even implemented them, but you
didn't want to introduce a dependency on Boost.  So you can't say I haven't
been bugging you constantly for seven years for this feature.  That's a lot
longer than eight months since I bugged you the last time.

**Comment by peastman** [comment]
I'm considering what the API for this should be.  On the Python side, I think the right thing to do is implement `__copy__()` and `__deepcopy__()`.  Both of them would do the same thing: a shallow copy of a System is never appropriate.

On the C++ side, the most idiomatic thing would be copy constructors, but that isn't compatible with how serialization works.  deserialize() creates a new object on the heap.  It can't be called to initialize an already existing object from inside that object's constructor.  An alternative is to create a static clone() method that can copy any object that has a registered serialization proxy.  Does that sound reasonable?

**Comment by jchodera** [comment]
Is a shallow copy really never appropriate?  Could there be some uses of shallow copy that are just fine, while incurring the overhead of a deepcopy might make them painful?  I'm just not sure here.

Regarding `clone()`: Is the idea that you want to avoid implementing copy constructors for the whole `System` and `Force` tree?

**Comment by peastman** [comment]
> Is a shallow copy really never appropriate?

Never.  When you add a Force or VirtualSite to a System, it takes ownership of that object.  If two Systems both had references to the same object, that would produce serious errors.

> Is the idea that you want to avoid implementing copy constructors for the whole System and Force tree?

It's that I *can't* implement them using serialization.  See above.  There's nothing I could put inside the constructor that would produce the correct result.

**Comment by jlmaccal** [comment]
Given the potentially slow performance of copying (even after optimization) and the complex issues surrounding object ownership, it seems like explicit clone method might be most appropriate.

**Comment by peastman** [comment]
Fixed by #1055.
